### ISSUE TITLE
test(amdgpu): add the Qwen3.8 HIP prefill device-parity and Vulkan A/B witness

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12598]
# Add the Qwen3.8 HIP prefill device-parity and Vulkan A/B witness

## Context

- For: Strix Halo maintainers deciding whether the #12539 HIP prefill panel is correct and worth retaining.
- Problem: #12539 currently has only a lexical source test; the advertised selector and Vulkan/HIP ablation tokens do not exist, and no test invokes the flat ABI on physical tensors.
- Today: `qwen38-long-prefill` is absent from the Strix selector/parity registry, `--ablate=vulkan,hip` is invalid, and the three-file candidate cannot emit numeric, timing, or continuation evidence.
- Better because: one source-bound device test can own both arms, typed component parity, repeated timings, and the exact keep/reject decision without inventing unsupported ablation tokens.
- Witness: pp2048 and pp32768 execute on physical gfx1151 for one warm-up plus at least five measured Vulkan/HIP repetitions per length, with a typed receipt.

- Centrality: Core
- P1 Context: advanced — binds the largest proposed AMD prefill lever to the exact Qwen3.8 attention geometry and physical candidate bytes.
- P2 Net value: advanced — measures both arms under one producer, preventing incomparable timing or hidden transfer/fallback accounting.
- P3 Adaptation: preserved — limits acceptance to a component panel; full-model handoff and continuation remain owned by #12540.
- P4 Operations: advanced — source, compiler, device, lease, timeout, memory, and per-run outcomes are captured together and failures stay typed.

## Current state

No registered selector, device test, or A/B receipt producer exists. The recovered #12539 kernel also requires correctness/performance repair before a physical spend is admissible.

## Parent context

Parent: #11572

## Dependencies and pickup rule

- Start blocked by: #12597
- Coordinates with: #12539
- Coordinates with: #12540
- Promotion requires: #12098

Pickup becomes executable when #12597 provides the generic ROCm Strix build route. It may then develop against the source-bound #12539 candidate without requiring that candidate to make an unwitnessed hardware claim first. #12098 gates external comparison claims only.

## Scope

Add one registered `qwen38-long-prefill` Strix selector and typed parity contract whose device test invokes the #12539 HIP ABI and the existing Vulkan component reference on identical resident F32 panels. The test owns both arms; the validator command therefore uses `--ablate=none`. Record allocation/setup separately from the measured panel, per-run and median prompt throughput, max-absolute output delta, finite outputs, kernel/transfer/fallback counters, and the exact geometry.

Excluded: full-model logits, KV-cache handoff, greedy continuation, HIP-to-Vulkan ownership transfer, external competitor comparisons, or a general ROCm backend.

## Core through-line

Register one selector -> compile the source-bound HIP candidate -> run identical resident component panels through HIP and Vulkan -> validate typed output/counter parity -> record the repeated keep/reject receipt.

## Gold-plating boundary

No full-model or serving integration, state handoff, competitor baseline, new backend router, or claim beyond the measured attention component.

## Done condition

On physical gfx1151, pp2048 and pp32768 each complete one warm-up plus at least five measured runs per Vulkan/HIP arm. Every run reports `engine=fak-native`, the correct backend, zero measured transfers and fallbacks, finite output, and component-output parity. The keep gate is HIP median at least 1.05x Vulkan at one length with no worse than 0.98x at the other; otherwise the receipt records REJECT and Vulkan remains selected.

## Witness

```text
fak-dev amd-strix-validate --subkernels=qwen38-long-prefill --ablate=none --evidence-only --json
```

The exact source/compiler/workload identity and independently approved timeout/abort budget must be recorded before the 24-trial physical run. A source change invalidates the review and receipt.

## Likely files

- `internal/amdgpu/strix_subkernels.go`
- `internal/amdgpu/strix_subkernels_test.go`
- `internal/compute/rocm_qwen38_prefill_device_test.go`

## Expected steps

5

## Work estimate

Estimate: 3 points.

## Overall completion contribution

Contribution: 3/100 points.

## Completion standard

experiment

## Target operating envelope

- physical gfx1151 executions: >= 24 runs
- prompt lengths covered: >= 2 lengths
- measured repetitions per arm and length: >= 5 runs
- component parity pass rate: = 100 percent
- finite output pass rate: = 100 percent
- measured transfer count: = 0 transfers
- backend fallback count: = 0 calls
- best-length HIP to Vulkan median speed ratio: >= 1.05 ratio
- other-length HIP to Vulkan median speed ratio: >= 0.98 ratio

## Witnessed operating envelope

- physical gfx1151 executions: = 0 runs
- prompt lengths covered: = 0 lengths
- measured repetitions per arm and length: = 0 runs
- component parity pass rate: = 0 percent
- finite output pass rate: = 0 percent
- measured transfer count: = 0 transfers
- backend fallback count: = 0 calls
- best-length HIP to Vulkan median speed ratio: = 0 ratio
- other-length HIP to Vulkan median speed ratio: = 0 ratio

### Work unit

leaf

## Closure binding

One signed, witnessed commit cites this issue and carries `(fak amdgpu)`; only a real gfx1151 receipt can close it.

**Comment by anthony-chaudhary** [comment]
Claimed by the dedicated #12598 task in isolated c926 worktree. Inspecting #12597 ownership and generic ROCm build prerequisite before implementation. This task owns the device parity/A-B witness, not #12539 kernel source. No physical window acquired; exact candidate source and independent budget review remain required.

**Comment by anthony-chaudhary** [comment]
Software WIP progressed: strict supplemental A/B receipt parser and regression, plus a draft physical device test invoking the allocation-free HIP Into helper and the existing synchronous resident Vulkan panel. The draft records setup separately, alternating warmup plus five measured samples per arm at pp2048/pp32768, counters, output comparison and derived verdict. It is not registered, tagged-compiled, or physically executed. No KEEP/REJECT measurement or completion claim is made. Remaining work: finish prerequisite #12597 source-bound build/controller integration, freeze the matching #12539 source, independently bind numerical tolerance and timeout/memory budget, then obtain sanctioned whole-device authority before the physical run. Current work is preserved in the dedicated isolated checkout; no commit or closure.

**Comment by anthony-chaudhary** [comment]
Software verification update: full internal/amdgpu tests passed in an isolated Linux clone at base f1bcc34e0a4962231d1ee5d72364c22accd2c1c7 plus the five source/test changes (actual go test exit 0, 381.956s). The subsequent source/compiler cross-binding addition passed focused parser/build-plan tests (actual exit 0, 0.023s). The sixth file, the draft device test, is excluded from this untagged run and remains uncompiled with the required ROCm/Vulkan tags. No physical witness or landing claim follows from these software results.

**Comment by anthony-chaudhary** [comment]
Source-owner coordination: the additive rocm_qwen38_prefill tag is accepted and implemented in #12539. Dual builder/device test should freeze with vulkan,rocm,rocm_qwen38_prefill. Final observed PCI BDF identity method PrefillPhysicalDeviceIdentity() returns validated canonical domain:bus:device.function; PrefillD2DBytesObserved() explicitly returns false. Runtime D2D is unsupported, not zero measured traffic; four-counter launch ABI is unchanged. Source bytes frozen in the existing candidate; final independent review/witness receipt is being published on #12539. No hardware authorization or commit implied.

**Comment by anthony-chaudhary** [comment]
Independent review corrections are software-tested: parser now binds both physical adapter identities to the reviewed selector, uses truthful host-transfer bytes rather than mixing counts and bytes, requires separately observed device-copy bytes, validates readback lengths in the draft producer, and compares review digest/source/compiler/device/timeout/numeric limit at the consumer. Vulkan launch is mutex-protected and requires one observed dispatch plus submit. Missing unique native device identity and HIP device-copy telemetry remain explicit HOLD conditions; no zero values are fabricated. Focused WSL regression including real temporary-artifact admission/mutation refusal passed with actual exit 0 (0.331s). Native seams and tagged/device execution remain pending under coordinated ownership; no completion or performance claim.

**Comment by anthony-chaudhary** [comment]
Confirmed complete: #12539 requested native updates were already implemented and independently verified. Final receipt: https://github.com/anthony-chaudhary/fak/issues/12539#issuecomment-5646931675 . Current de97 source unchanged: rocm.go bc5af5ed92e35000df645d75e56d1195f5a0bbab8689c55333fbb076f60abba4; HIP 41974e5b82c1256362e1e68e47c90a945c37233093b35d34bbf5631a8ddcccdd; source test aeab6c3f5d6bc23e41cefedcd70da246e4981c37d2fa87c6c0ead84c6c25c8b2. PrefillPhysicalDeviceIdentity() returns canonical actual HIP PCI BDF; PrefillD2DBytesObserved() returns false (explicit unsupported instrumentation, never manufactured zero). Extra rocm_qwen38_prefill tag is implemented; dual build/device test use vulkan,rocm,rocm_qwen38_prefill. Independent source review and direct coordinator WSL witness passed actual exit0 (0.041s). No source work outstanding for this request; consume these frozen bytes read-only. No hardware/commit permission conveyed.

**Comment by anthony-chaudhary** [comment]
Dependency clarification with #12540: this issue owns only resident F32 attention-component Q/K/V/output A/B. Even a future KEEP receipt here will not establish ownership of real model attention-KV or GDN state, exported handles/semaphores, no-replay handoff, or continuation parity. Coordinates with #12540 remains non-blocking coordination; this component is not its full-state-owner prerequisite. The missing full-state owner and handoff acceptance remain durable/open under #12540 (see its prerequisite review comment 5646981278). Current physical acceptance remains absent.

**Comment by anthony-chaudhary** [comment]
Shared native Vulkan PCI identity edits remain held by unresolved authoritative leases. Lease-recovery work #12877 has a verified repair candidate, but its mandatory validation is blocked by #6827 and it has not landed or mutated production refs. Existing #12390/#12396 remote refs therefore remain uncleared; this task will not use the unlanded repair or override those refs. Next dependency step: resolve the gate, validate/land #12877, then perform fenced owner recovery and acquire the exact Vulkan identity paths. Component software WIP stays preserved; no physical or landing claim.
